File: android/src/main/java/com/facebook/flipper/testing/FlipperConnectionMock.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.flipper.testing;

import com.facebook.flipper.core.FlipperArray;
import com.facebook.flipper.core.FlipperConnection;
import com.facebook.flipper.core.FlipperObject;
import com.facebook.flipper.core.FlipperReceiver;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FlipperConnectionMock implements FlipperConnection {
  public final Map<String, FlipperReceiver> receivers = new HashMap<>();
  public final Map<String, List<Object>> sent = new HashMap<>();
  public final List<Throwable> errors = new ArrayList<>();

  @Override
  public void send(String method, FlipperObject params) {
    final List<Object> paramList;
    if (sent.containsKey(method)) {
      paramList = sent.get(method);
    } else {
      paramList = new ArrayList<>();
      sent.put(method, paramList);
    }

    paramList.add(params);
  }

  @Override
  public void send(String method, FlipperArray params) {
    final List<Object> paramList;
    if (sent.containsKey(method)) {
      paramList = sent.get(method);
    } else {
      paramList = new ArrayList<>();
      sent.put(method, paramList);
    }

    paramList.add(params);
  }

  @Override
  public void reportErrorWithMetadata(String reason, String stackTrace) {
    errors.add(new Throwable(reason));
  }

  @Override
  public void reportError(Throwable throwable) {
    errors.add(throwable);
  }

  @Override
  public void receive(String method, FlipperReceiver receiver) {
    receivers.put(method, receiver);
  }
}

File: iOS/FlipperKitTestUtils/FlipperConnectionMock.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FlipperConnectionMock.h"

@implementation FlipperConnectionMock

- (instancetype)init {
  if (self = [super init]) {
    _connected = YES;
    _receivers = @{};
    _sent = @{};
    _sentWithArray = @{};
  }
  return self;
}

- (void)sendInternal:(NSString*)method
          withParams:(id)params
            loggedTo:(NSDictionary* __strong*)sentLog {
  if (_connected) {
    NSMutableDictionary* newSentLog = [NSMutableDictionary new];
    [newSentLog addEntriesFromDictionary:*sentLog];
    if (newSentLog[method]) {
      newSentLog[method] = [(*sentLog)[method] arrayByAddingObject:params];
    } else {
      newSentLog[method] = @[ params ];
    }
    *sentLog = newSentLog;
  }
}

- (void)send:(NSString*)method withParams:(NSDictionary*)params {
  [self sendInternal:method withParams:params loggedTo:&_sent];
}

- (void)send:(NSString*)method withRawParams:(NSString*)params {
  [self sendInternal:method withParams:params loggedTo:&_sent];
}

- (void)send:(NSString*)method withArrayParams:(NSArray*)params {
  [self sendInternal:method withParams:params loggedTo:&_sentWithArray];
}

- (void)receive:(NSString*)method withBlock:(SonarReceiver)receiver {
  if (_connected) {
    NSMutableDictionary* newReceivers = [NSMutableDictionary new];
    [newReceivers addEntriesFromDictionary:_receivers];
    newReceivers[method] = receiver;
    _receivers = newReceivers;
  }
}

- (void)errorWithMessage:(NSString*)message stackTrace:(NSString*)stacktrace {
  // Empty Implementation
  // TODO: Test this method too.
}

@end
